internal/sapsystem/sapinstance.go

```diff
@@ -0,0 +1,138 @@
+	"github.com/trento-project/agent/internal/sapsystem/sapcontrolapi"
+	"github.com/trento-project/agent/internal/utils"
+)
+
+var (
+	databaseFeatures         = regexp.MustCompile("HDB.*")
+	applicationFeatures      = regexp.MustCompile("MESSAGESERVER.*|ENQREP|ABAP.*")
+	diagnosticsAgentFeatures = regexp.MustCompile("SMDAGENT")
+)
+
+type SystemReplication map[string]interface{}
+type HostConfiguration map[string]interface{}
+type HdbnsutilSRstate map[string]interface{}
+
+type SAPInstance struct {
+	Name       string
+	Type       SystemType
+	Host       string
+	SAPControl *SAPControl
+	// Only for Database type
+	SystemReplication SystemReplication
+	HostConfiguration HostConfiguration
+	HdbnsutilSRstate  HdbnsutilSRstate
+}
+
+func NewSAPInstance(w sapcontrolapi.WebService, executor utils.CommandExecutor) (*SAPInstance, error) {
+	host, err := os.Hostname()
+	if err != nil {
+		return nil, err
+	}
+
+	scontrol, err := NewSAPControl(w)
+	if err != nil {
+		return nil, err
+	}
+
+	instanceName, err := scontrol.findProperty("INSTANCE_NAME")
+	if err != nil {
+		return nil, err
+	}
+
+	instanceType, err := detectType(scontrol)
+	if err != nil {
+		return nil, err
+	}
+
+	sapInstance := &SAPInstance{
+		Host:              host,
+		SAPControl:        scontrol,
+		Name:              instanceName,
+		Type:              instanceType,
+		SystemReplication: nil,
+		HostConfiguration: nil,
+		HdbnsutilSRstate:  nil,
+	}
+
+	if instanceType == Database {
+		sid, err := sapInstance.SAPControl.findProperty("SAPSYSTEMNAME")
+		if err != nil {
+			return nil, errors.Wrap(err, "Error finding the SAP instance sid")
+		}
+
+		sapInstance.SystemReplication = systemReplicationStatus(executor, sid, sapInstance.Name)
+		sapInstance.HostConfiguration = landscapeHostConfiguration(executor, sid, sapInstance.Name)
+		sapInstance.HdbnsutilSRstate = hdbnsutilSrstate(executor, sid, sapInstance.Name)
+	}
+
+	return sapInstance, nil
+}
+
+func detectType(sapControl *SAPControl) (SystemType, error) {
+	sapLocalhost, err := sapControl.findProperty("SAPLOCALHOST")
+	if err != nil {
+		return Unknown, err
+	}
+
+	for _, instance := range sapControl.Instances {
+		if instance.Hostname == sapLocalhost {
+			switch {
+			case databaseFeatures.MatchString(instance.Features):
+				return Database, nil
+			case applicationFeatures.MatchString(instance.Features):
+				return Application, nil
+			case diagnosticsAgentFeatures.MatchString(instance.Features):
+				return DiagnosticsAgent, nil
+			default:
+				return Unknown, nil
+			}
+		}
+	}
+
+	return Unknown, nil
+}
+
+func runPythonSupport(executor utils.CommandExecutor, sid, instance, script string) map[string]interface{} {
+	user := fmt.Sprintf("%sadm", strings.ToLower(sid))
+	cmdPath := path.Join(sapInstallationPath, sid, instance, "exe/python_support", script)
+	cmd := fmt.Sprintf("python %s --sapcontrol=1", cmdPath)
+	// Even with a error return code, some data is available
+	srData, err := executor.Exec("su", "-lc", cmd, user)
```

Comment:
The error is checked in the next condition and printed. If it doesn't return an expected list of `key=value` outputs it will return an empty map, which is what we need to return in this case to populate the struct fields